balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+        }
+
+        // Step 2. Mix buffer contents.
+        // for t from 0 to t_cost-1:
+        for t in 0..t_cost {
+            // for m from 0 to s_cost-1:
+            for m in 0..s_cost {
+                // Step 2a. Hash last and current blocks.
+                // block_t prev = buf[(m-1) mod s_cost]
+                let prev = if m == 0 { s_cost - 1 } else { m - 1 };
+                let prev = &memory_blocks[prev];
+
+                // buf[m] = hash(cnt++, prev, buf[m])
+                digest.update(&cnt.to_le_bytes());
+                cnt = cnt.wrapping_add(1);
+                digest.update(prev);
+                digest.update(&memory_blocks[m]);
+                memory_blocks[m] = digest.finalize_reset();
+
+                // Step 2b. Hash in pseudorandomly chosen blocks.
+                // for i from 0 to delta-1:
+                for i in 0..DELTA {
+                    // block_t idx_block = ints_to_block(t, m, i)
+                    // int other = to_int(hash(cnt++, salt, idx_block)) mod s_cost
+                    digest.update(&cnt.to_le_bytes());
+                    cnt = cnt.wrapping_add(1);
+                    digest.update(salt);
+                    digest.update(&u32::try_from(t).unwrap().to_le_bytes());
+                    digest.update(&u32::try_from(m).unwrap().to_le_bytes());
+                    digest.update(&i.to_le_bytes());
+                    let other =
+                        u32::from_le_bytes(digest.finalize_reset()[..4].try_into().unwrap())
+                            % self.params.s_cost.get();
+                    let other = usize::try_from(other).unwrap();
+
+                    // buf[m] = hash(cnt++, buf[m], buf[other])
+                    digest.update(&cnt.to_le_bytes());
+                    cnt = cnt.wrapping_add(1);
+                    digest.update(&memory_blocks[m]);
+                    digest.update(&memory_blocks[other]);
+                    memory_blocks[m] = digest.finalize_reset();
+                }
+            }
+        }
+
+        // Step 3. Extract output from buffer.
+        // return buf[s_cost-1]
+        Ok(memory_blocks[s_cost - 1].clone())
+    }
+}
+
+#[cfg(all(feature = "alloc", feature = "password-hash"))]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+#[cfg_attr(docsrs, doc(cfg(feature = "password-hash")))]
+impl<D: Digest> PasswordHasher for Balloon<'_, D> {
+    type Params = Params;
+
+    fn hash_password<'a, S>(
+        &self,
+        password: &[u8],
+        salt: &'a S,
+    ) -> password_hash::Result<PasswordHash<'a>>
+    where
+        S: AsRef<str> + ?Sized,
+    {
+        let salt = Salt::try_from(salt.as_ref())?;
+        let mut salt_arr = [0u8; 64];
+        let salt_bytes = salt.b64_decode(&mut salt_arr)?;
+
+        let output = password_hash::Output::new(&self.hash(password, salt_bytes)?)?;
+
+        Ok(PasswordHash {
+            algorithm: Ident::new("balloon"),
+            version: Some(1),
```

Comment:
We don't have a standard or spec, it's "just" a research paper, it never mentions PHC strings. I just took those from the reference implementation.